src/query/ovfp_threahold_monitor.cpp

```diff
@@ -0,0 +1,626 @@
+// class ovfp_threshold
+// private
+void
+ovfp_threshold::clear (INDEX_OVFP_INFO *root)
+{
+  INDEX_OVFP_INFO *pt;
+  while (root)
+    {
+      pt = root;
+      root = pt->next;
+      free (pt);
+    }
+}
+
+void
+ovfp_threshold::free_info_mem (INDEX_OVFP_INFO *pt)
+{
+  assert (pt != NULL);
+  pt->next = m_free_head ? m_free_head : NULL;
+  m_free_head = pt;
+}
+
+INDEX_OVFP_INFO *
+ovfp_threshold::add (BTID *btid, OID *class_oid, time_t time_now, int npages)
+{
+  INDEX_OVFP_INFO  *pt;
+
+  pt = find (btid, class_oid);
+  if (pt)
+    {
+      pt->event_time[RECENT_POS] = time_now;
+      pt->read_pages[RECENT_POS] = npages;
+
+      if (pt->read_pages[MAX_POS] <= npages)
+	{
+	  pt->event_time[MAX_POS] = time_now;
+	  pt->read_pages[MAX_POS] = npages;
+	}
+      pt->hit_cnt++;
+
+      return pt;
+    }
+
+
+  pt = alloc_info_mem();
+  assert (pt != NULL);
+  if (pt)
+    {
+      // new item
+      BTID_COPY ( & (pt->btid), btid);
+      COPY_OID (& (pt->class_oid), class_oid);
+
+      pt->event_time[RECENT_POS] = time_now;
+      pt->read_pages[RECENT_POS] = npages;
+      pt->event_time[MAX_POS] = time_now;
+      pt->read_pages[MAX_POS] = npages;
+      pt->hit_cnt = 1;
+
+      if (m_head == NULL || m_prev == NULL)
+	{
+	  pt->next = m_head;
+	  m_head = pt;
+	}
+      else
+	{
+	  pt->next = m_prev->next;
+	  m_prev->next = pt;
+	}
+    }
+
+  return pt;
+}
+
+//protected
+INDEX_OVFP_INFO  *
+ovfp_threshold::alloc_info_mem()
+{
+  INDEX_OVFP_INFO *pt;
+
+  if (m_free_head)
+    {
+      pt = m_free_head;
+      m_free_head = m_free_head->next;
+    }
+  else
+    {
+      pt = (INDEX_OVFP_INFO *) malloc (sizeof (INDEX_OVFP_INFO));
+      assert (pt != NULL);
+    }
+  return pt;
+}
+
+INDEX_OVFP_INFO *
+ovfp_threshold::find (BTID *btid, OID *class_oid)
+{
+  int cmp;
+
+  if (m_head == NULL)
+    {
+      m_prev = NULL;
+      return NULL;
+    }
+
+  for (INDEX_OVFP_INFO *cur = m_head; cur; cur = cur->next)
+    {
+      cmp = OID_COMPARE (& (cur->class_oid), class_oid);
+      if ( cmp == 0 )
+	{
+	  cmp = BTID_COMPARE (& (cur->btid), btid);
+	}
+
+      if ( cmp > 0 )
+	{
+	  return NULL;
+	}
+      else if ( cmp == 0 )
+	{
+	  return cur;
+	}
+
+      m_prev = cur;
+    }
+
+  return NULL;
+}
+
+//public
+ovfp_threshold::ovfp_threshold()
+{
+  m_free_head = m_head = m_prev = NULL;
+
+  m_lock_idx = -1;
+  m_lock_ptr = NULL;
+}
+ovfp_threshold::~ovfp_threshold()
+{
+  if (m_lock_ptr)
+    {
+      m_lock_ptr->lock (m_lock_idx, m_lock_idx);
+      clear (m_head);
+      clear (m_free_head);
+      m_lock_ptr->unlock (m_lock_idx, m_lock_idx);
+    }
+  else
+    {
+      clear (m_head);
+      clear (m_free_head);
+    }
+  m_head = m_free_head = NULL;
+}
+
+void
+ovfp_threshold::set_worker_idx (int idx, ovfp_monitor_lock *lock_mgr_p)
+{
+  assert (idx >= 0 && idx <= VACUUM_MAX_WORKER_COUNT);
+  m_lock_idx = idx;
+  if (m_lock_idx >= 0 && m_lock_idx < VACUUM_MAX_WORKER_COUNT)
```

Comment:
The assert check was done right above that.
Although the if statement is unnecessary because it is after the assert check,
Assert doesn't work in release, so it's added just in case.